clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberAssignmentReconciler.java

```diff
@@ -0,0 +1,291 @@
+                                           PartitionDiffer partitionDiffer,
+                                           EventGenerator eventGenerator,
+                                           AssignmentModifier assignmentModifier) {
+        // Check for any outstanding operations first. If a conclusive result has already been reached, return that
+        // before processing any further.
+        Optional<ReconciliationResult> inflightStatus = checkInflightStatus(timer);
+
+        if (inflightStatus.isPresent())
+            return inflightStatus.get();
+
+        // "diff" the two sets of partitions: our "current" assignment and the "target" assignment. The result is
+        // sorted primarily so when the partitions show up in the logs, it's easier for us humans to understand.
+        Set<TopicPartition> diffPartitions = sortPartitions(partitionDiffer.diff(subscriptions.assignedPartitions(),
+                targetPartitions(metadata, assignment)));
+
+        // TODO: should this be here? Do we expect the above to do anything besides move things around in memory?
+        timer.update();
+
+        if (diffPartitions.isEmpty()) {
+            log.debug("No partitions were {} in the new assignment", operation.verbPastTense);
+            return ReconciliationResult.NO_CHANGE;
+        }
+
+        // Set up our callback invocation. We don't block here waiting on it its completion, though.
+        log.debug("Preparing to invoke {} with the following partitions: {}", operation.methodName, diffPartitions);
+        RebalanceCallbackEvent event = eventGenerator.generate(Collections.unmodifiableList(new ArrayList<>(diffPartitions)));
+        inflightCallback = Optional.of(event);
+
+        // Enqueue it in our background->application shared queue. This should be invoked in the Consumer.poll() method.
+        backgroundEventQueue.add(event);
+
+        // When the core event is completed in the application via Consumer.poll(), we want to invoke our code to
+        // "commit" the change.
+        event.future().whenComplete((result, error) -> {
+            if (error != null) {
+                log.warn("An error occurred when invoking {} with the following partitions: {}", operation.methodName, diffPartitions, error);
+                // TODO: should we proceed or abort?
+            }
+
+            log.debug("{} was successfully executed with the following partitions: {}", operation.methodName, diffPartitions);
+
+            // Make a copy of the current assignment set (just to be paranoid) and then modify it. The "modification"
+            // here will be different for a revocation vs. an assignment.
+            Set<TopicPartition> newAssignment = new HashSet<>(subscriptions.assignedPartitions());
+            assignmentModifier.modify(newAssignment, diffPartitions);
+
+            // And finally... assign the new set of partitions!
+            subscriptions.assignFromSubscribed(newAssignment);
```

Comment:
I would add a comment pointing out that the subscription can only be updated after the callbacks are executed so we make sure that user will be able to commit offsets in the callbacks on revocation. 

We have the same behaviour for assignment/revocation but it's truly the revocation+commit what forces this dependency between the callback exec and making the subscription effective (and start fetching)